libraries/lib-mixer/Mix.cpp

```diff
@@ -78,6 +96,214 @@ size_t FindBufferSize(
+   virtual size_t NChannels() const = 0;
+   //! Gain multiplier that should be applied to the channel
+   virtual float GetChannelGain(size_t channel) const = 0;
+   //! For the given `iChannel` fills the channelFlags array, that describes
+   //! to which output it should go.
+   virtual void FindChannelFlags(unsigned char *channelFlags, size_t numChannels, size_t iChannel) = 0;
+   //! Returns true if source channels could be combined into mono if needed
+   virtual bool CanMakeMono() const = 0;
+};
+
+class SequenceDownmixSource final : public DownmixSource
+{
+   AudioGraph::Source& mDownstream;
+   const WideSampleSequence& mSequence;
+   //! many-to-one mixing of channels
+   //! Pointer into array of arrays
+   const ArrayOf<bool> * mpMap {};
+public:
+
+   SequenceDownmixSource(AudioGraph::Source& downstream,
+                         const WideSampleSequence& sequence,
+                         //! Null or else must have a lifetime enclosing this objects's
+                         const ArrayOf<bool> *channelMap)
+      : mDownstream(downstream)
+      , mSequence(sequence)
+      , mpMap(channelMap)
+   {
+   }
+
+   AudioGraph::Source& GetDownstream() const override { return mDownstream; }
+   size_t NChannels() const override { return mSequence.NChannels(); }
+   float GetChannelGain(size_t channel) const override { return mSequence.GetChannelGain(channel); }
+   void FindChannelFlags(unsigned char *channelFlags, size_t numChannels, size_t iChannel) override
+   {
+      const bool* map = mpMap ? mpMap[iChannel].get() : nullptr;
+      const auto end = channelFlags + numChannels;
+      std::fill(channelFlags, end, 0);
+      if (map)
+         // ignore left and right when downmixing is customized
+         std::copy(map, map + numChannels, channelFlags);
+      else if (AudioGraph::IsMono(mSequence))
+         std::fill(channelFlags, end, 1);
+      else if (iChannel == 0)
+         channelFlags[0] = 1;
+      else if (iChannel == 1) {
+         if (numChannels >= 2)
+            channelFlags[1] = 1;
+         else
+            channelFlags[0] = 1;
+      }
+   }
+   bool CanMakeMono() const override { return mpMap == nullptr; }
+};
+
+class StereoToMono final : public DownmixSource
+{
+   AudioGraph::Source& mMasterSource;
+public:
+
+   StereoToMono(AudioGraph::Source& masterSource)
+      : mMasterSource(masterSource)
+   { }
+
+   AudioGraph::Source& GetDownstream() const override { return mMasterSource; }
+   size_t NChannels() const override { return 2; }
+   float GetChannelGain(size_t) const override { return 1.f; }
+   void FindChannelFlags(unsigned char* channelFlags, size_t numChannels, size_t) override
+   {
+      for(size_t i = 0; i < numChannels; ++i)
+         channelFlags[i] = true;
+   }
+   bool CanMakeMono() const override { return true; }
+};
+
+//! Combines multiple audio graph sources into a single source
+class DownmixStage final : public AudioGraph::Source
```

Comment:
I like the initiative of moving some of the complexity behind appropriate abstractions. Before this gets merged, though, I'd like to see these new classes get their own files. Then functions like `MixBuffers` would also be relocated to `DownmixStage.cpp`, and maybe others, too.